src/common/bip352.h

```diff
@@ -0,0 +1,169 @@
+// Copyright (c) 2023 The Bitcoin Core developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#ifndef BITCOIN_COMMON_BIP352_H
+#define BITCOIN_COMMON_BIP352_H
+
+#include <key.h>
+#include <pubkey.h>
+#include <primitives/transaction.h>
+#include <secp256k1.h>
+#include <secp256k1_silentpayments.h>
+#include <uint256.h>
+#include <addresstype.h>
+#include <coins.h>
+
+#include <vector>
+#include <variant>
+
+namespace BIP352 {
+
+using PubKey = std::variant<CPubKey, XOnlyPubKey>;
+using PubTweakData = secp256k1_silentpayments_public_data;
+
+struct BIP352Comparator {
+    bool operator()(const COutPoint& a, const COutPoint& b) const {
+        // BIP352 defines the "smallest outpoint" based on a lexicographic
+        // sort of the outpoints, using the 36-byte serialization:
+        // <txid, 32-bytes little-endian>:<vout, 4-bytes little-endian>
+        if (a.hash != b.hash) {
+            return a.hash < b.hash;
+        }
+        uint8_t a_vout[4], b_vout[4];
+        WriteLE32(a_vout, a.n);
+        WriteLE32(b_vout, b.n);
+        return std::memcmp(a_vout, b_vout, 4) < 0;
```

Comment:
You don't need to do this. You can just compare the numbers as unsigned ints in memory. It will yield identical results on all platforms and be faster.

Reply:
Actually wait nevermind.. the thing is defined as lexically sorted versus the le32 serialization of the int. Forget I said anything -- you do need to do this.